### PR TITLE
Fix bug for inconsistent lists served from etcd

### PR DESCRIPTION
**Pull request by wojtek-t**
```release-note
kube-apiserver: fixed a bug returning inconsistent results from list requests which set a field or label selector and set a paging limit
```

/kind bug
/priority critical-urgent

### CODE REVIEW

**Comment by wojtek-t** [comment]
@deads2k @smarterclayton - in GKE we have watchcache enabled for everything but events, but I'm wondering why you've never hit this problem given you have watchcache disabled ...

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go [hunk 1]
```diff
@@ -695,6 +695,10 @@ func (s *store) List(ctx context.Context, key string, opts storage.ListOptions,
 			break
 		}
 		key = string(lastKey) + "\x00"
+		if fromRV == nil {
+			*fromRV = uint64(returnedRV)
+			options = append(options, clientv3.WithRev(returnedRV))
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:700]: appending to options inside this for loop looks wrong... that means we'll have multiple WithRev options specified with different values

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:700]: edit: nevermind, I see this only happens when fromRV is nil

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go [hunk 2]
```diff
@@ -695,6 +695,10 @@ func (s *store) List(ctx context.Context, key string, opts storage.ListOptions,
 			break
 		}
 		key = string(lastKey) + "\x00"
+		if fromRV == nil {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:698]: if we're using ResourceVersionMatchNotOlderThan, a non-nil fromRV doesn't mean we added a WithRev option.

I think we should track whether we added WithRev explicitly, and make that the condition here.

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:698]: That also means we need a test that demonstrates paging+filtering+ResourceVersionMatchNotOlderThan that would exercise this

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:698]: Good catch, test added and test is already catching it.

Now I need to fix the code.

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:698]: Fixed

**Comment by wojtek-t** [comment]
@liggitt - thanks for review - PTAL

**Comment by liggitt** [comment]
can you open cherry-picks to 1.16-1.18 for this?

**Comment by wojtek-t** [comment]
> can you open cherry-picks to 1.16-1.18 for this?

Will do on Monday, since these aren't automatic ones unfortunately....

**Comment by wojtek-t** [comment]
Cherrypicks are opened.

**Pull request merged**